### PR TITLE
Separate Alonzo genesis file

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/LedgerState.hs [hunk 1]
```diff
@@ -442,6 +445,9 @@ data NodeConfig = NodeConfig
   , ncByronGenesisHash :: !GenesisHashByron
   , ncShelleyGenesisFile :: !GenesisFile
   , ncShelleyGenesisHash :: !GenesisHashShelley
+  -- TODO alonzo: This fields should be used
```
Comment (erikd) [inline cardano-api/src/Cardano/Api/LedgerState.hs:448]: These?

Comment (newhoggy) [reply cardano-api/src/Cardano/Api/LedgerState.hs]: Fixed.  Thanks!

File: configuration/chairman/byron-shelley/configuration.yaml [hunk 2]
```diff
@@ -8,6 +8,7 @@
 
 ByronGenesisFile: byron/genesis.json
 ShelleyGenesisFile: shelley/genesis.json
+AlonzoGenesisFile: alonzo/genesis.json
```
Comment (newhoggy) [inline configuration/chairman/byron-shelley/configuration.yaml:11]: New entry in the `configuration.yaml`

File: scripts/byron-to-alonzo/mkfiles.sh [hunk 3]
```diff
@@ -319,12 +320,41 @@ echo
...
+echo "Generated shelley/genesis.json:"
 echo
 cat shelley/genesis.json
 echo
 echo "====================================================================="
 
+# TODO alonzo: Should this be generated by cardano-cli genesis create?
+
+mkdir -p alonzo
+cat > alonzo/genesis.json <<EOF
```
Comment (newhoggy) [inline scripts/byron-to-alonzo/mkfiles.sh:332]: Also need a new `alonzo/genesis.json` file.

File: cardano-node/src/Cardano/Node/Configuration/POM.hs [hunk 4]
```diff
@@ -242,6 +243,24 @@ instance FromJSON PartialNodeConfiguration where
              , npcShelleyGenesisFileHash
              }
 
+      parseAlonzoProtocol v = do
+        primary   <- v .:? "AlonzoGenesisFile"
+        secondary <- v .:? "GenesisFile"
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Configuration/POM.hs:248]: I made this mistake too. Apparently we don't need the `"GenesisFile"` field here as it was for backwards compatibility reasons.

File: scripts/byron-to-alonzo/mkfiles.sh [hunk 5]
```diff
@@ -319,12 +320,41 @@ echo
...
 echo "====================================================================="
 
-echo "Generated genesis.json:"
+echo "Generated shelley/genesis.json:"
 echo
 cat shelley/genesis.json
 echo
 echo "====================================================================="
 
+# TODO alonzo: Should this be generated by cardano-cli genesis create?
```
Comment (Jimbo4350) [inline scripts/byron-to-alonzo/mkfiles.sh:329]: So I think we should modify `cardano-cli genesis create` to also create an alonzo genesis file. It shouldn't be too much of a change as most of the right bits are already there.

File: scripts/byron-to-alonzo/mkfiles.sh [hunk 6]
```diff
@@ -319,12 +320,41 @@ echo
...
+{
+  "adaPerUTxOWord": 0,
+  "executionPrices": {
+    "prMem": 1,
+    "prSteps": 1
+  },
+  "maxTxExUnits": {
+    "exUnitsMem": 1,
+    "exUnitsSteps": 1
+  },
```
Comment (Jimbo4350) [inline scripts/byron-to-alonzo/mkfiles.sh:342]: Once `cardano-cli genesis create` is modified to create an alonzo genesis file, we can remove this.

File: scripts/byron-to-alonzo/mkfiles.sh [hunk 7]
```diff
@@ -560,10 +590,6 @@ if [ "$1" = "alonzo" ]; then
   sed -i ${ROOT}/configuration.yaml \
       -e 's/LastKnownBlockVersion-Major: 1/LastKnownBlockVersion-Major: 5/'
 
-  # Update shelley genesis with required Alonzo fields.
-  alonzogenesisparams='.+ {adaPerUTxOWord: 0, executionPrices: {prMem: 1, prSteps: 1}, maxTxExUnits: {exUnitsMem: 1, exUnitsSteps: 1}, maxBlockExUnits: {exUnitsMem: 1, exUnitsSteps: 1}, maxValueSize: 1000, collateralPercentage: 100, maxCollateralInputs: 1}'
```
Comment (Jimbo4350) [inline scripts/byron-to-alonzo/mkfiles.sh:564]: :+1:

**Review by Jimbo4350 — COMMENTED** [review]
Looking good! We need to incorporate the Alonzo genesis hash check similarly to:
https://github.com/input-output-hk/cardano-node/blob/ad76ebe63e138f5bdaa91117f7583b145b687b85/cardano-node/src/Cardano/Node/Protocol/Shelley.hs#L107

And then update the mainnet configuration files in the node.

File: configuration/chairman/byron-shelley/configuration.yaml (hunk 2 above)
Comment (disassembler) [reply configuration/chairman/byron-shelley/configuration.yaml:11]: If we include a new genesis file shouldn't we also include a hash in that file?

File: cardano-node/src/Cardano/Node/Configuration/POM.hs (hunk 4 above)
Comment (newhoggy) [reply cardano-node/src/Cardano/Node/Configuration/POM.hs]: `GenesisFile` is removed.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs [hunk 8]
```diff
@@ -366,6 +366,26 @@ runGenesisCreate (GenesisDir rootdir)
...
+      , "exUnitsSteps" .= Aeson.toJSON @Int 1
+      ]
+    , "maxBlockExUnits" .= Aeson.object
+      [ "exUnitsMem" .= Aeson.toJSON @Int 1
+      , "exUnitsSteps" .= Aeson.toJSON @Int 1
+      ]
+    , "maxValueSize" .= Aeson.toJSON @Int 1000
+    , "collateralPercentage" .= Aeson.toJSON @Int 100
+    , "maxCollateralInputs" .= Aeson.toJSON @Int 1
+    ]
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:388]: `cardano-cli genesis create` now creates `<rootdir>/genesis.alonzo.json`.

File: cardano-api/src/Cardano/Api/LedgerState.hs [hunk 9]
```diff
@@ -745,6 +764,37 @@ renderShelleyGenesisError sge =
...
+          , "configuration file is ", renderHash expected, "."
+          ]
+
+      AlonzoGenesisDecodeError fp err ->
+        mconcat
+          [ "There was an error parsing the genesis file: ", Text.pack fp
+          , " Error: ", err
+          ]
+  where
+    renderHash :: GenesisHashAlonzo -> Text
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/LedgerState.hs:795]: `cardano-binary` exposes `hashToBytesAsHex`

File: scripts/byron-to-alonzo/mkfiles.sh (hunk 6 above)
Comment (newhoggy) [reply scripts/byron-to-alonzo/mkfiles.sh]: Removed!

File: scripts/byron-to-alonzo/mkfiles.sh (hunk 5 above)
Comment (newhoggy) [reply scripts/byron-to-alonzo/mkfiles.sh]: Done!

File: scripts/byron-to-alonzo/mkfiles.sh (hunk 3 above)
Comment (newhoggy) [reply scripts/byron-to-alonzo/mkfiles.sh]: This is now `shelley/genesis.alonzo.json`

File: configuration/chairman/byron-shelley/configuration.yaml (hunk 2 above)
Comment (Jimbo4350) [reply configuration/chairman/byron-shelley/configuration.yaml]: You can but it's not mandatory.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs (hunk 8 above)
Comment (jbgi) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:388]: should `create-staked` also do that?

**Review by dcoutts — APPROVED** [review]
All the comments below are now outdated as I've resolved them all.

**Pull request merged**